### PR TITLE
[SPARK-8129] [CORE] [Sec] Pass auth secrets to executors via env variables

### PR DESCRIPTION
**Pull request by kanzhang**
Env variables are not visible to non-Spark users, based on suggestion from @vanzin.

### CODE REVIEW

**Comment by srowen** [comment]
@vanzin is this roughly what you had in mind?

**Comment by vanzin** [comment]
Yes, this is a lot more inline with what I was thinking before. Thanks for making the change.

One thing I noticed is that `SparkConf.isExecutorStartupConf` is unchanged. It should not include the secret in the list of options to pass on an executor's command line.

(Note I'll be out for about 10 days so I won't be able to follow up on reviews.)

**Comment by kanzhang** [comment]
> One thing I noticed is that SparkConf.isExecutorStartupConf is unchanged. It should not include the secret in the list of options to pass on an executor's command line.

Thanks for catching it. Agree it shouldn't be passed in the first place, even though the actual filtering in the patch will remove it later. Uploaded a fix.

**Comment by kanzhang** [comment]
> Yes, this is a lot more inline with what I was thinking before. Thanks for making the change.

Actually, this is the kind of patch I wanted to "preempt" with my previous patch, because I think eventually when we support per-app secrets in standalone mode, auth secrets should be set earlier when we build `ApplicationDescription` and `DriverDescription`. Oh well, will have to try that later.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala [hunk 1]
```diff
@@ -365,10 +365,12 @@ private[spark] class SecurityManager(sparkConf: SparkConf)
       cookie
     } else {
       // user must have set spark.authenticate.secret config
-      sparkConf.getOption("spark.authenticate.secret") match {
+      // For Master/Worker, auth secret is in conf; for Executors, it is in env variable
+      sys.env.get(SecurityManager.ENV_AUTH_SECRET)
+        .orElse(sparkConf.getOption(SecurityManager.SPARK_AUTH_SECRET_CONF)) match {
         case Some(value) => value
         case None => throw new Exception("Error: a secret key must be specified via the " +
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: This isn't your code, but if we have to change this PR again, let's throw something better than `Exception` here.

**Comment by vanzin** [comment]
> Actually, this is the kind of patch I wanted to "preempt" with my previous patch, because I think eventually when we support per-app secrets in standalone mode

And as I said numerous times, whether to use env variables or something else to propagate a secret has absolutely zero to do with whether you can support per-app secrets.

**Comment by srowen** [comment]
Sounds like violent agreement. If there are no more comments I'll get this merged.

File: core/src/main/scala/org/apache/spark/SecurityManager.scala (hunk 1 above)
Comment (kanzhang) [inline core/src/main/scala/org/apache/spark/SecurityManager.scala]: SparkException?

**Comment by kanzhang** [comment]
> And as I said numerous times, whether to use env variables or something else to propagate a secret has absolutely zero to do with whether you can support per-app secrets.

No disagreement on that statement. My comment only concerns the implementation aspects of it. If stdin is a viable approach and we agree to use it for per-app secret support now, it gives me justification to introduce `appSecret` in `ApplicationDescription` and `DriverDescription` and handle all filtering and setting up app secrets when we generate them, which I think is the way to go when we add per-app secret support, even if we continue to use the current env variable approach. No need to do the same at a lower-level like `buildLocalCommand`, as it is now.

**Comment by vanzin** [comment]
>  If stdin is a viable approach and we agree to use it for per-app secret support now, it gives me justification to introduce appSecret in ApplicationDescription and DriverDescription

That's where I don't follow you. If you want to add per-app secrets, you have that justification regardless of how that secret is propagated. But this change is not about per-app secrets, it's about how the secret is currently passed in the command line.

**Comment by kanzhang** [comment]
> If you want to add per-app secrets, you have that justification regardless of how that secret is propagated.

I can't add it if it doesn't offer any added security benefit. Currently, with env variable and jobs not run as user, there is no added security to go per-app secrets.

> But this change is not about per-app secrets, it's about how the secret is currently passed in the command line.

Agree per-app secrets is better done in a separate patch. Let's not hold up this one.

**Comment by vanzin** [comment]
> with env variable and jobs not run as user, there is no added security to go per-app secrets.

You can just say "all jobs run as the same user" there. That's the main impediment to having proper security in standalone. With that out of the way, the env variable is a non-issue. Which is what I've been telling you all along.

**Comment by kanzhang** [comment]
And I get what you are saying all along. :)

**Comment by kanzhang** [comment]
It is becoming clear to me auth secret should be properly set (and filtered out) when we create `ApplicationDescription` and `DriverDescription` (as shown in my previous patch), regardless how we pass the secret to child processes. Although it involves more changes (in the 3 places where we create those descriptions), it is preferable since it allows us to set things correct early. Not intended to make any change now, just a note for future reference.

**Pull request closed without merging**